src/node/operator_event.rs

```diff
@@ -121,45 +133,114 @@ fn resolve_access_conflicts(x: &OperatorEvent, y: &OperatorEvent) -> Ordering {
 /// Ordering used in the lattice where `self < other` implies `self` *precedes* other.
 impl Ord for OperatorEvent {
     fn cmp(&self, other: &OperatorEvent) -> Ordering {
-        match (self.is_watermark_callback, other.is_watermark_callback) {
-            (true, true) => {
-                // Both of the events are watermarks, so the watermark with the lower timestamp
-                // should run first. Ties are broken by first by dependencies to ensure
-                // that writes to state occur before reads, and then by priority where a smaller
-                // number is higher priority.
-                match self.timestamp.cmp(&other.timestamp) {
-                    Ordering::Equal => match resolve_access_conflicts(&self, other) {
-                        // Prioritize other.
-                        Ordering::Equal => Ordering::Greater,
-                        ord => ord,
-                    },
-                    ord => ord,
+        match self.operator_type {
+            OperatorType::ReadOnly => {
+                match (self.is_watermark_callback, other.is_watermark_callback) {
+                    (true, true) => {
+                        // Both of the events are watermarks, so the watermark with the lower
+                        // timestamp should run first. Ties are broken by first by dependencies to
+                        // ensure that writes to state occur before reads, and then by priority
+                        // where a smaller number is higher priority.
+                        match self.timestamp.cmp(&other.timestamp) {
+                            Ordering::Equal => match resolve_access_conflicts(&self, other) {
```

Comment:
Can this result in `unreachable!()`? Read-only operators should never run 2 watermarks for the same timestamp.

Reply:
I think this is old code from before we had strictly defined interfaces for the operators. Our current redesign should ensure a single watermark callback for both "ReadOnly" and "Writeable" operators. Cleaning this and addressing the comments below.